test/transport_end2end_test.go

```diff
@@ -0,0 +1,198 @@
+// transport and the old transport is closed.
+func (s) TestClientTransportRestartsAfterStreamIDExhausted(t *testing.T) {
+	// Set the transport's MaxStreamID to 5 to cause connection to drain after 2 RPCs.
+	originalMaxStreamID := transport.MaxStreamID
+	transport.MaxStreamID = 5
+	defer func() {
+		transport.MaxStreamID = originalMaxStreamID
+	}()
+
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for i := 0; i < 2; i++ {
+				if _, err := stream.Recv(); err != nil {
+					return status.Errorf(codes.Internal, "unexpected error receiving: %v", err)
+				}
+				if err := stream.Send(&testpb.StreamingOutputCallResponse{}); err != nil {
+					return status.Errorf(codes.Internal, "unexpected error sending: %v", err)
+				}
+			}
+			if recv, err := stream.Recv(); err != io.EOF {
+				return status.Errorf(codes.Internal, "Recv = %v, %v; want _, io.EOF", recv, err)
+			}
+			return nil
+		},
+	}
+
+	creds := &transportRestartCheckCreds{}
+	if err := ss.Start(nil, grpc.WithTransportCredentials(creds)); err != nil {
+		t.Fatalf("Starting stubServer: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	var streams []testpb.TestService_FullDuplexCallClient
+
+	// expectedNumConns when each stream is created.
+	expectedNumConns := []int{1, 1, 2}
+
+	// Set up 3 streams and call sendAndReceive() once on each.
+	for i := 0; i < 3; i++ {
+		s, err := ss.Client.FullDuplexCall(ctx)
+		if err != nil {
+			t.Fatalf("Creating FullDuplex stream: %v", err)
+		}
+
+		streams = append(streams, s)
+		if err := s.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+			t.Fatalf("Sending on stream %d: %v", i, err)
+		}
+		if _, err := s.Recv(); err != nil {
+			t.Fatalf("Receiving on stream %d: %v", i, err)
+		}
+
+		// Verify expected num of conns.
+		if len(creds.connections) != expectedNumConns[i] {
+			t.Fatalf("Number of connections created: %v, want: %v", len(creds.connections), expectedNumConns[i])
+		}
+	}
+
+	// Verify all streams still work.
+	for i, stream := range streams {
+		if err := stream.Send(&testpb.StreamingOutputCallRequest{}); err != nil {
+			t.Fatalf("Sending on stream %d: %v", i, err)
+		}
+		if _, err := stream.Recv(); err != nil {
+			t.Fatalf("Receiving on stream %d: %v", i, err)
+		}
+	}
+
+	var connPerStream []net.Conn
+
+	// The peer passed via the call option is set up only after the RPC is complete.
+	// Conn used by the stream is available in authInfo.
+	for i, stream := range streams {
+		if err := stream.CloseSend(); err != nil {
+			t.Fatalf("CloseSend() on stream %d: %v", i, err)
+		}
+		p, ok := peer.FromContext(stream.Context())
+		if !ok {
+			t.Fatalf("Getting peer from stream context for stream %d", i)
+		}
+		connPerStream = append(connPerStream, p.AuthInfo.(*authInfoWithConn).conn)
+	}
+
+	// Verifying the first and second RPCs were made on the same connection.
```

Comment:
I'm not even sure we need any of this anymore, since you're verifying the number of connections after each stream, above.  Please consider and simplify as much as possible (without compromising the validity/coverage of the test).

Reply:
you are correct. This doesnt seem to be important anymore. I can also get rid of the AuthInfo stuff along with this.